### PR TITLE
Adding a callback function register to improve flexibility of Keras

### PR DESCRIPTION
**Pull request by qinwang-ai**
Hello everyone, I added one member function in `DirectoryIterator` class allow developer to register callback functions when pumping data by `flow_from_directory` of `ImageDataGenerator` class. It's can improve flexibility when have complex data structure of feature and label. It's very useful to register a customize function to process their own logic to integrate the `batch_x` and `batch_y`. It can make the `DirectoryIterator` pump the data structure what the model required.

> example:

```
def process_train_batch_X(raw_x_batch, filenames, index_array):
    tmp_temperature = []
    tmp_time = []
    for i, j in enumerate(index_array):
        tmp_temperature.append(get_temp_from_str(filenames[j]))
        tmp_time.append(get_time_from_str(filenames[j]))
    # get discretization array from the filename
    x_batch_temperature = utils.discretization(tmp_temperature)
    x_batch_time = utils.discretization(tmp_time)
    x_batch = [raw_x_batch, x_batch_temperature, x_batch_time]
    return x_batch

def process_batch_Y(raw_x_batch, filenames, index_array):
    .....

train_iterator.register_batch_processor(process_train_batch_X, process_batch_Y)
```

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by fchollet** [comment]
Thank you for the PR.

`preprocessing_function` is an existing argument to `ImageDataGenerator`, which allows you to add a custom processing step for each batch of data. It is called at the start of `standardize`.

In you need more advanced types of preprocessing than what can be covered by this built-in setup, I recommend that you subclass `ImageDataGenerator` to add this functionality while reusing as much of the built-in features as possible.

After reviewing your proposal, we decided not to merge it.

By the way, this part of the code has migrated to its own repository: https://github.com/keras-team/keras-preprocessing

In the future, if you wish to make any PRs to Keras preprocessing, please send your PR to this repository.
